### PR TITLE
Add debug-only range checks

### PR DESCRIPTION
**Pull request by AdamNorberg**
We're writing code assuming that we're operating on well-defined unique numeric values - our math isn't designed to work around NaN or infinity. Checking for them can add up, so let's add checks to debug mode only.

### CODE REVIEW

File: engine/geometry/Cylinder.cs [hunk 1]
```diff
@@ -210,6 +213,11 @@ public override dvec3 GetPositionAt(dvec2 uv)
 		/// <inheritdoc />
 		public override List<Vertex> GenerateVertexList(int resolutionU, int resolutionV)
 		{
+			// If asked to sample at zero points, return an empty list.
+			if (resolutionU <= 0 || resolutionV <= 0)
+			{
+				return new List<Vertex>();
+			}
```
Comment (Lathreas) [inline engine/geometry/Cylinder.cs:220]: Good check!

File: engine/calculus/QuarticFunction.cs [hunk 2]
```diff
@@ -77,45 +82,40 @@ public IEnumerable<double> Roots()
...
+				foreach (double v in CubicFunction.Solve(c0, c1, c2, c3))
 				{
 					yield return v;
 				}
 				yield break;
 			}
+
+			double ba = c3/c4;
+			double ca = c2/c4;
+			double da = c1/c4;
```
Comment (Lathreas) [inline engine/calculus/QuarticFunction.cs:103]: Thanks for keeping these explicit! They're important for numeric precision, and that would change if they would be inlined. So this is good.

**Review by Lathreas — COMMENTED** [review]
Well designed class DebugUtil, I like it! There are two minor style changes needed (DebugUtil class has spaces instead of tabs for indentation, which the other files don't have, and I think a small naming change to QuarticFunction.Solve would make it more consistent). After those are done this can be merged IMO.

File: engine/utilities/DebugUtil.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        /// <param name="items">Collection of items to range check.</param>
+        /// <param name="varName">Name of variable being checked (use nameof); included in error message.</param>
+        [Conditional("DEBUG")]
+        public static void AssertAllFinite(SortedPointsList<double> items, string varName)
+        {
+            AssertAllFinite(items.Key, $"{varName}.Key");
+            AssertAllFinite(items.Value, $"{varName}.Value");
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (AdamNorberg) [reply engine/utilities/DebugUtil.cs:78]: Whoops, fixed my IDE settings and reformatted.

File: engine/calculus/QuarticFunction.cs [hunk 4]
```diff
@@ -77,45 +82,40 @@ public IEnumerable<double> Roots()
 		}
 		
 		/// <summary>
-		///     Solves the equation \f$e + d x + c x^2 + b x^3 + a x^4 = 0\f$, returning all real values of
+		///     Solves the equation \f$c0 + c1 x + c2 x^2 + c3 x^3 + c4 x^4 = 0\f$, returning all real values of
 		///		\f$x\f$ for which the equation is true. See https://en.wikipedia.org/wiki/Quartic_function for the
 		///		algorithm used.
 		/// </summary>
-		public static IEnumerable<double> Solve(double e2, double d2, double c2, double b2, double a2)
+		public static IEnumerable<double> Solve(double c0, double c1, double c2, double c3, double c4)
```
[the remark this answers is not in the document]
Comment (AdamNorberg) [reply engine/calculus/QuarticFunction.cs]: Done.

**Comment by AdamNorberg** [comment]
Done. Thanks for the quick review!

**Pull request merged**